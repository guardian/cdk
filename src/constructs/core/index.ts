export * from "./identity";
export * from "./migrating";
export * from "./parameters";
export * from "./ssm";
export * from "./stack";
