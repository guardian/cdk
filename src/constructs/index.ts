export * as acm from "./acm";
export * as autoscaling from "./autoscaling";
export * as cloudwatch from "./cloudwatch";
export * as core from "./core";
export * as dns from "./dns";
export * as dynamodb from "./dynamodb";
export * as ec2 from "./ec2";
export * as ecs from "./ecs";
export * as iam from "./iam";
export * as kinesis from "./kinesis";
export * as lambda from "./lambda";
export * as loadbalancing from "./loadbalancing";
export * as rds from "./rds";
export * as s3 from "./s3";
export * as ses from "./ses";
export * as vpc from "./vpc";
export * as root from "./root";
