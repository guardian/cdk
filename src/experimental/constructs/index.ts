export * from "./error-budget-alarm";
export * from "./root";
