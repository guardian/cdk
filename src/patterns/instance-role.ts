import { ServicePrincipal } from "@aws-cdk/aws-iam";
import type { GuStack } from "../constructs/core";
import type { GuGetS3ObjectPolicyProps, GuLogShippingPolicyProps, GuPolicy } from "../constructs/iam";
import {
  GuDescribeEC2Policy,
  GuGetS3ObjectPolicy,
  GuLogShippingPolicy,
  GuRole,
  GuSSMRunCommandPolicy,
} from "../constructs/iam";

interface InstanceRoleProps extends GuGetS3ObjectPolicyProps, Partial<GuLogShippingPolicyProps> {
  additionalPolicies?: GuPolicy[];
}

export class InstanceRole extends GuRole {
  private policies: GuPolicy[];

  constructor(scope: GuStack, props: InstanceRoleProps) {
    super(scope, "InstanceRole", {
      overrideId: true,
      path: "/",
      assumedBy: new ServicePrincipal("ec2.amazonaws.com"),
    });

    this.policies = [
      new GuSSMRunCommandPolicy(scope),
      new GuGetS3ObjectPolicy(scope, "GetDistributablesPolicy", props),
      new GuDescribeEC2Policy(scope),
      ...(props.loggingStreamName
        ? [new GuLogShippingPolicy(scope, "LogShippingPolicy", props as GuLogShippingPolicyProps)]
        : []),
      ...(props.additionalPolicies ? props.additionalPolicies : []),
    ];

    this.policies.forEach((p) => p.attachToRole(this));
  }
}
