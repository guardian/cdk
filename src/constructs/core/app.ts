import { Tags } from "aws-cdk-lib";
import { Construct } from "constructs";
import type { Identity } from "./identity";
import type { GuStack } from "./stack";

export class GuApp extends Construct implements Identity {
  public readonly stack: string;
  public readonly stage: string;
  public readonly app: string;

  /**
   * A sanitised version of `app` suitable to use for the logical ID of CloudFormation resources, which must be alphanumeric.
   * Typically, only needed when wanting to create a static (deterministic) logical ID during synthesis.
   *
   * @see https://stackoverflow.com/a/20864946
   */
  public readonly appForLogicalId: string;

  public readonly parent: GuStack;

  // eslint-disable-next-line custom-rules/valid-constructors -- `scope` cannot be `GuApp` as that's recursive
  constructor(scope: GuStack, id: string) {
    super(scope, id);

    this.stack = scope.stack;
    this.stage = scope.stage;
    this.app = id;

    const titleCaseApp = id.charAt(0).toUpperCase() + id.slice(1);
    this.appForLogicalId = titleCaseApp.replace(/[\W_]+/g, "");

    this.parent = scope;

    Tags.of(this).add("App", id);
  }
}
