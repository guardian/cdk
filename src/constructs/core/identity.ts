import type { IConstruct } from "@aws-cdk/core";
import { Tags } from "@aws-cdk/core";

export interface StackStageIdentity {
  stack: string;
  stage: string;
}

export interface AppIdentity {
  app: string;
}

export interface Identity extends StackStageIdentity, AppIdentity {}

export const AppIdentity = {
  suffixText(appIdentity: AppIdentity, text: string): string {
    const titleCaseApp = appIdentity.app.charAt(0).toUpperCase() + appIdentity.app.slice(1);
    return `${text}${titleCaseApp}`;
  },
  taggedConstruct<T extends IConstruct>(appIdentity: AppIdentity, construct: T): T {
    Tags.of(construct).add("App", appIdentity.app);
    return construct;
  },
};
