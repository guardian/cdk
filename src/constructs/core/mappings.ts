import { CfnMapping } from "@aws-cdk/core";
import type { Stage } from "../../constants";
import type { AppIdentity } from "./identity";
import type { GuStack } from "./stack";

export type GuMappingValue = string | number | boolean;

export interface GuStageMappingValue<T extends GuMappingValue> extends AppIdentity {
  variableName: string;
  stageValues: Record<Stage, T>;
}

export class GuStageMapping {
  private readonly scope: GuStack;

  // the key in this Map is an app string
  private mappings: Map<string, CfnMapping>;

  constructor(scope: GuStack) {
    this.scope = scope;
    this.mappings = new Map<string, CfnMapping>();
  }

  private getCfnMapping(app: string): CfnMapping {
    if (!this.mappings.has(app)) {
      const emptyCfnMapping = new CfnMapping(this.scope, app);
      this.mappings.set(app, emptyCfnMapping);
      return emptyCfnMapping;
    }

    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion -- `this.mappings.has` above is `true`, so we can safely assert non-null
    return this.mappings.get(app)!;
  }

  withValue<T extends GuMappingValue>(mappingValue: GuStageMappingValue<T>): T {
    const { app, variableName, stageValues } = mappingValue;

    const cfnMapping = this.getCfnMapping(app);

    for (const [stage, value] of Object.entries(stageValues)) {
      cfnMapping.setValue(stage, variableName, value);
    }

    return cfnMapping.findInMap(this.scope.stage, variableName) as T;
  }
}
