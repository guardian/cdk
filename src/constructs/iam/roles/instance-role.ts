import { ServicePrincipal } from "@aws-cdk/aws-iam";
import { AppIdentity } from "../../core/identity";
import {
  GuDescribeEC2Policy,
  GuGetDistributablePolicy,
  GuLogShippingPolicy,
  GuParameterStoreReadPolicy,
  GuSSMRunCommandPolicy,
} from "../policies";
import { GuRole } from "./roles";
import type { GuStack } from "../../core";
import type { GuPolicy } from "../policies";

export interface GuInstanceRoleProps {
  withoutLogShipping?: boolean;
  additionalPolicies?: GuPolicy[];
}

export type GuInstanceRolePropsWithApp = GuInstanceRoleProps & AppIdentity;

/**
 * Creates an IAM role with common policies that are needed by most Guardian applications.
 *
 * More specifically:
 * 1. Allows for `ssh` access to an EC2 instance via [ssm-scala](https://github.com/guardian/ssm-scala) (instead of standard `ssh`).
 * 2. Allows EC2 instances to communicate with Wazuh, for security monitoring.
 * 3. Allows EC2 instances to download an artifact from AWS S3, for application deployment.
 * 4. Allows EC2 instances to download private configuration from AWS Parameter Store.
 * 5. Allows EC2 instances to write logs into our central ELK stack via Kinesis.
 *
 * If additional IAM permissions are required, create custom policies and pass them in via the `additionalPolicies` prop.
 *
 * If log shipping is not required, opt out by setting the `withoutLogShipping` prop to `true`.
 */
export class GuInstanceRole extends GuRole {
  constructor(scope: GuStack, props: GuInstanceRolePropsWithApp) {
    super(scope, AppIdentity.suffixText(props, "InstanceRole"), {
      path: "/",
      assumedBy: new ServicePrincipal("ec2.amazonaws.com"),
      // not setting existingLogicalId results in the logicalId always being auto-generated
    });

    const sharedPolicies = [
      GuSSMRunCommandPolicy.getInstance(scope),
      GuDescribeEC2Policy.getInstance(scope),
      ...(props.withoutLogShipping ? [] : [GuLogShippingPolicy.getInstance(scope)]),
    ];

    const policies = [
      ...sharedPolicies,
      new GuGetDistributablePolicy(scope, props),
      new GuParameterStoreReadPolicy(scope, props),
      ...(props.additionalPolicies ? props.additionalPolicies : []),
    ];

    policies.forEach((p) => p.attachToRole(this));

    AppIdentity.taggedConstruct(props, this);
  }
}
