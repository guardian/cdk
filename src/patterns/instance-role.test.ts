import "@aws-cdk/assert/jest";
import { SynthUtils } from "@aws-cdk/assert";
import { App } from "@aws-cdk/core";
import { GuStack } from "../constructs/core";
import { GuGetS3ObjectPolicy } from "../constructs/iam";
import { InstanceRole } from "./instance-role";

describe("The InstanceRole construct", () => {
  it("should create the correct resources with minimal config", () => {
    const stack = new GuStack(new App());
    new InstanceRole(stack, { bucketName: "test" });

    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
    expect(stack).toCountResources("AWS::IAM::Role", 1);
    expect(stack).toCountResources("AWS::IAM::Policy", 3);
  });

  it("should create an additional logging policy if logging stream is specified", () => {
    const stack = new GuStack(new App());
    new InstanceRole(stack, { bucketName: "test", loggingStreamName: "my-logging-stream" });

    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
    expect(stack).toCountResources("AWS::IAM::Role", 1);
    expect(stack).toCountResources("AWS::IAM::Policy", 4);
  });

  it("should allow additional policies to be specified", () => {
    const stack = new GuStack(new App());

    new InstanceRole(stack, {
      bucketName: "test",
      additionalPolicies: [new GuGetS3ObjectPolicy(stack, "GetConfigPolicy", { bucketName: "config" })],
    });

    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
    expect(stack).toCountResources("AWS::IAM::Role", 1);
    expect(stack).toCountResources("AWS::IAM::Policy", 4);
  });
});
