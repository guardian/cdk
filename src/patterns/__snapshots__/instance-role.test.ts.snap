// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The InstanceRole construct should allow additional policies to be specified 1`] = `
Object {
  "Parameters": Object {
    "Stack": Object {
      "Default": "deploy",
      "Description": "Name of this stack",
      "Type": "String",
    },
    "Stage": Object {
      "AllowedValues": Array [
        "CODE",
        "PROD",
      ],
      "Default": "CODE",
      "Description": "Stage name",
      "Type": "String",
    },
  },
  "Resources": Object {
    "DescribeEC2PolicyFF5F9295": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "autoscaling:DescribeAutoScalingInstances",
                "autoscaling:DescribeAutoScalingGroups",
                "ec2:DescribeTags",
                "ec2:DescribeInstances",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "describe-ec2-policy",
        "Roles": Array [
          Object {
            "Ref": "InstanceRole",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "GetConfigPolicy6F934A1C": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Resource": "arn:aws:s3:::config/*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GetConfigPolicy6F934A1C",
        "Roles": Array [
          Object {
            "Ref": "InstanceRole",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "GetDistributablesPolicy648EC38C": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Resource": "arn:aws:s3:::test/*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GetDistributablesPolicy648EC38C",
        "Roles": Array [
          Object {
            "Ref": "InstanceRole",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "InstanceRole": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "ec2.",
                      Object {
                        "Ref": "AWS::URLSuffix",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Path": "/",
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "testing",
          },
          Object {
            "Key": "Stack",
            "Value": Object {
              "Ref": "Stack",
            },
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ParameterStoreRead9D2F4FAB": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:ssm:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/",
                    Object {
                      "Ref": "Stage",
                    },
                    "/",
                    Object {
                      "Ref": "Stack",
                    },
                    "/testing",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "parameter-store-read-policy",
        "Roles": Array [
          Object {
            "Ref": "InstanceRole",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "SSMRunCommandPolicy244E1613": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "ec2messages:AcknowledgeMessage",
                "ec2messages:DeleteMessage",
                "ec2messages:FailMessage",
                "ec2messages:GetEndpoint",
                "ec2messages:GetMessages",
                "ec2messages:SendReply",
                "ssm:UpdateInstanceInformation",
                "ssm:ListInstanceAssociations",
                "ssm:DescribeInstanceProperties",
                "ssm:DescribeDocumentParameters",
                "ssmmessages:CreateControlChannel",
                "ssmmessages:CreateDataChannel",
                "ssmmessages:OpenControlChannel",
                "ssmmessages:OpenDataChannel",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ssm-run-command-policy",
        "Roles": Array [
          Object {
            "Ref": "InstanceRole",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;

exports[`The InstanceRole construct should create an additional logging policy if logging stream is specified 1`] = `
Object {
  "Parameters": Object {
    "LoggingStreamName": Object {
      "Default": "/account/services/logging.stream.name",
      "Description": "SSM parameter containing the Name (not ARN) on the kinesis stream",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "Stack": Object {
      "Default": "deploy",
      "Description": "Name of this stack",
      "Type": "String",
    },
    "Stage": Object {
      "AllowedValues": Array [
        "CODE",
        "PROD",
      ],
      "Default": "CODE",
      "Description": "Stage name",
      "Type": "String",
    },
  },
  "Resources": Object {
    "DescribeEC2PolicyFF5F9295": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "autoscaling:DescribeAutoScalingInstances",
                "autoscaling:DescribeAutoScalingGroups",
                "ec2:DescribeTags",
                "ec2:DescribeInstances",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "describe-ec2-policy",
        "Roles": Array [
          Object {
            "Ref": "InstanceRole",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "GetDistributablesPolicy648EC38C": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Resource": "arn:aws:s3:::test/*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GetDistributablesPolicy648EC38C",
        "Roles": Array [
          Object {
            "Ref": "InstanceRole",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "GuLogShippingPolicy981BFE5A": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "kinesis:Describe*",
                "kinesis:Put*",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:kinesis:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":stream/",
                    Object {
                      "Ref": "LoggingStreamName",
                    },
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GuLogShippingPolicy981BFE5A",
        "Roles": Array [
          Object {
            "Ref": "InstanceRole",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "InstanceRole": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "ec2.",
                      Object {
                        "Ref": "AWS::URLSuffix",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Path": "/",
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "testing",
          },
          Object {
            "Key": "Stack",
            "Value": Object {
              "Ref": "Stack",
            },
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ParameterStoreRead9D2F4FAB": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:ssm:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/",
                    Object {
                      "Ref": "Stage",
                    },
                    "/",
                    Object {
                      "Ref": "Stack",
                    },
                    "/testing",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "parameter-store-read-policy",
        "Roles": Array [
          Object {
            "Ref": "InstanceRole",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "SSMRunCommandPolicy244E1613": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "ec2messages:AcknowledgeMessage",
                "ec2messages:DeleteMessage",
                "ec2messages:FailMessage",
                "ec2messages:GetEndpoint",
                "ec2messages:GetMessages",
                "ec2messages:SendReply",
                "ssm:UpdateInstanceInformation",
                "ssm:ListInstanceAssociations",
                "ssm:DescribeInstanceProperties",
                "ssm:DescribeDocumentParameters",
                "ssmmessages:CreateControlChannel",
                "ssmmessages:CreateDataChannel",
                "ssmmessages:OpenControlChannel",
                "ssmmessages:OpenDataChannel",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ssm-run-command-policy",
        "Roles": Array [
          Object {
            "Ref": "InstanceRole",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;

exports[`The InstanceRole construct should create the correct resources with minimal config 1`] = `
Object {
  "Parameters": Object {
    "Stack": Object {
      "Default": "deploy",
      "Description": "Name of this stack",
      "Type": "String",
    },
    "Stage": Object {
      "AllowedValues": Array [
        "CODE",
        "PROD",
      ],
      "Default": "CODE",
      "Description": "Stage name",
      "Type": "String",
    },
  },
  "Resources": Object {
    "DescribeEC2PolicyFF5F9295": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "autoscaling:DescribeAutoScalingInstances",
                "autoscaling:DescribeAutoScalingGroups",
                "ec2:DescribeTags",
                "ec2:DescribeInstances",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "describe-ec2-policy",
        "Roles": Array [
          Object {
            "Ref": "InstanceRole",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "GetDistributablesPolicy648EC38C": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Resource": "arn:aws:s3:::test/*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GetDistributablesPolicy648EC38C",
        "Roles": Array [
          Object {
            "Ref": "InstanceRole",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "InstanceRole": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "ec2.",
                      Object {
                        "Ref": "AWS::URLSuffix",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Path": "/",
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "testing",
          },
          Object {
            "Key": "Stack",
            "Value": Object {
              "Ref": "Stack",
            },
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ParameterStoreRead9D2F4FAB": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:ssm:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/",
                    Object {
                      "Ref": "Stage",
                    },
                    "/",
                    Object {
                      "Ref": "Stack",
                    },
                    "/testing",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "parameter-store-read-policy",
        "Roles": Array [
          Object {
            "Ref": "InstanceRole",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "SSMRunCommandPolicy244E1613": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "ec2messages:AcknowledgeMessage",
                "ec2messages:DeleteMessage",
                "ec2messages:FailMessage",
                "ec2messages:GetEndpoint",
                "ec2messages:GetMessages",
                "ec2messages:SendReply",
                "ssm:UpdateInstanceInformation",
                "ssm:ListInstanceAssociations",
                "ssm:DescribeInstanceProperties",
                "ssm:DescribeDocumentParameters",
                "ssmmessages:CreateControlChannel",
                "ssmmessages:CreateDataChannel",
                "ssmmessages:OpenControlChannel",
                "ssmmessages:OpenDataChannel",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ssm-run-command-policy",
        "Roles": Array [
          Object {
            "Ref": "InstanceRole",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
